File: backend/app/schemas/analysis.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.government_claim import GovernmentClaimResult
from app.schemas.threat_intelligence import ThreatIntelResult

SCHEMA_VERSION = "1.0"
HARD_MAX_BODY_CHARS = 20_000

ContentSource = Literal["pasted_text", "manual_entry", "ocr", "user_corrected_ocr", "url_input"]
Confidence = Literal["low", "medium", "high"]
RiskLevel = Literal["LOW", "MEDIUM", "HIGH", "CRITICAL"]


class _Strict(BaseModel):
    model_config = ConfigDict(extra="forbid")


# ---------- request ----------


class ContentIn(_Strict):
    body: str = Field(default="", max_length=HARD_MAX_BODY_CHARS)
    source: ContentSource
    user_confirmed: Literal[True] = Field(
        description="Must be true: the user reviewed this exact content and chose to submit it."
    )


class SenderIn(_Strict):
    value: str | None = Field(default=None, max_length=64)
    kind: Literal["phone_number", "alphanumeric_sender_id", "display_name", "unknown"] = "unknown"
    provenance: Literal["user_entered", "ocr", "user_corrected_ocr", "unavailable"] = "unavailable"
    # No auditable authority exists in this MVP, so nothing may be submitted as verified.
    verification_status: Literal["unverified"] = "unverified"


class AttachmentMeta(_Strict):
    type: Literal["screenshot"] = "screenshot"
    provenance: Literal["user_upload"] = "user_upload"
    original_filename: str | None = Field(default=None, max_length=255)
    declared_mime_type: str | None = Field(default=None, max_length=100)


class PrivacyIn(_Strict):
    upload_confirmed: bool
    retention_preference: Literal["delete_after_analysis"] = "delete_after_analysis"


class AnalyzeRequest(_Strict):
    schema_version: str = SCHEMA_VERSION
    content: ContentIn
    sender: SenderIn | None = None
    attachment: AttachmentMeta | None = None
    language_preference: Literal["en", "ta", "both"] = "both"
    privacy: PrivacyIn

    @model_validator(mode="after")
    def _body_required_unless_screenshot(self) -> "AnalyzeRequest":
        ocr_path = self.content.source in ("ocr", "user_corrected_ocr") or self.attachment is not None
        if not self.content.body.strip() and not ocr_path:
            raise ValueError("content.body must not be empty for text or URL input")
        if self.attachment is not None and not self.privacy.upload_confirmed:
            raise ValueError("privacy.upload_confirmed must be true when a screenshot is attached")
        return self


# ---------- response ----------


class RiskOut(BaseModel):
    assessment_status: Literal["assessed", "insufficient_content"] = Field(
        description="`insufficient_content` means there was no text to analyse; the level must not be read as safe."
    )
    level: RiskLevel
    score: int = Field(ge=0, le=100)
    score_label: Literal["indicator_strength"] = "indicator_strength"
    score_note: str = (
        "The score reflects the strength of detected risk indicators. It is not a probability "
        "that the message is a scam."
    )
    score_note_ta: str = (
        "இந்த மதிப்பெண் கண்டறியப்பட்ட ஆபத்துக் குறிகளின் வலிமையைக் காட்டுகிறது. இது செய்தி மோசடி என்பதற்கான "
        "நிகழ்தகவு அல்ல."
    )
    verdict_scope: str
    verdict_scope_ta: str | None = None


class EvidenceItem(BaseModel):
    signal: str
    category: Literal["message", "url", "threat_intelligence", "government_claim", "provenance"]
    source: str
    kind: Literal["risk", "info"] = "risk"
    confidence: Confidence
    rule_id: str | None = None
    description_en: str
    description_ta: str | None = None
    observed: str | None = Field(
        default=None, description="Short excerpt of what triggered the rule, echoed back only to this user"
    )


class DomainCheckOut(BaseModel):
    signal: str
    confidence: Confidence
    detail: str
    url: str


class UrlIntelligenceOut(BaseModel):
    extracted_urls: list[str]
    primary_url: str | None
    domain_checks: list[DomainCheckOut]
    provider_enabled: bool
    provider_result: ThreatIntelResult | None
    privacy_note: str | None = None


class ProvenanceOut(BaseModel):
    content_source: ContentSource
    verification_status: Literal["unverified"] = "unverified"
    user_confirmed: bool
    attachment_received: bool
    ocr_status: Literal["not_applicable", "not_available_in_this_build", "extracted", "failed"]
    character_count: int


class SenderAssessmentOut(BaseModel):
    value_masked: str | None
    kind: str
    provenance: str
    verification_status: Literal["unverified"] = "unverified"
    evidence_weight: Literal["low"] = "low"
    warnings: list[str]
    warnings_ta: list[str] = Field(default_factory=list)


class ExplanationOut(BaseModel):
    en: str
    ta: str
    generated_by: Literal["gemini", "template"]
    ai_status: Literal["generated", "disabled", "unavailable", "rejected"] = Field(
        default="disabled",
        description="`generated`: AI explanation passed validation. Otherwise the template explanation is shown.",
    )
    model: str | None = None
    note: str | None = None
    note_ta: str | None = None


class ProviderFlags(BaseModel):
    virustotal_enabled: bool
    virustotal_available: bool | None = Field(description="None when no URL was checked")
    gemini_enabled: bool
    gemini_available: bool | None = Field(description="None when no explanation call was attempted")


class AnalyzeResponse(BaseModel):
    schema_version: str = SCHEMA_VERSION
    analysis_id: str
    risk: RiskOut
    evidence: list[EvidenceItem]
    url_intelligence: UrlIntelligenceOut
    government_claim: GovernmentClaimResult
    provenance: ProvenanceOut
    sender_assessment: SenderAssessmentOut
    missing_metadata: list[str]
    limitations: list[str]
    limitations_ta: list[str] = Field(default_factory=list)
    safe_next_steps: list[str]
    safe_next_steps_ta: list[str]
    explanation: ExplanationOut
    provider_flags: ProviderFlags


class ErrorResponse(BaseModel):
    error: str
    detail: str
    analysis_id: str | None = None

File: frontend/src/types/analysis.ts
// Mirrors backend/app/schemas/analysis.py. Keep in sync with /openapi.json.

export type ContentSource = "pasted_text" | "manual_entry" | "ocr" | "user_corrected_ocr" | "url_input";
// Where the image behind screenshot text came from. The image itself is never sent.
export type ImageOrigin = "upload" | "camera";
export type Confidence = "low" | "medium" | "high";
export type RiskLevel = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";
export type Language = "en" | "ta";
export type AiStatus = "generated" | "disabled" | "unavailable" | "rejected";

export type SenderKind = "phone_number" | "alphanumeric_sender_id" | "display_name" | "unknown";

export interface AnalyzeRequest {
  schema_version: "1.0";
  content: { body: string; source: ContentSource; user_confirmed: true; image_origin?: ImageOrigin | null };
  sender?: {
    value: string | null;
    kind: SenderKind;
    provenance: "user_entered" | "ocr" | "user_corrected_ocr" | "unavailable";
    verification_status: "unverified";
  };
  attachment?: {
    type: "screenshot";
    provenance: "user_upload";
    original_filename?: string | null;
    declared_mime_type?: string | null;
  };
  language_preference: "en" | "ta" | "both";
  privacy: { upload_confirmed: boolean; retention_preference: "delete_after_analysis" };
}

export interface EvidenceItem {
  signal: string;
  category: "message" | "url" | "threat_intelligence" | "government_claim" | "provenance";
  source: string;
  kind: "risk" | "info";
  confidence: Confidence;
  rule_id: string | null;
  description_en: string;
  description_ta: string | null;
  observed: string | null;
}

export type ThreatIntelStatus =
  | "malicious"
  | "suspicious"
  | "not_found"
  | "clean_or_harmless"
  | "unavailable"
  | "unknown";

export interface ThreatIntelResult {
  provider: string;
  indicator_type: string;
  indicator: string;
  status: ThreatIntelStatus;
  available: boolean;
  unavailable_reason: string | null;
  engine_stats: { malicious: number; suspicious: number; harmless: number; undetected: number; timeout: number } | null;
  engines_total: number | null;
  categories: string[];
  source_timestamp: string | null;
  checked_at: string;
  from_cache: boolean;
  note: string;
  note_ta?: string | null;
}

export type GovernmentClaimStatus =
  | "supported_by_curated_kb"
  | "partially_supported_by_curated_kb"
  | "contradicted_by_curated_kb"
  | "not_found_in_curated_kb"
  | "not_government_related"
  | "unable_to_assess";

export interface ClaimField {
  value: string | null;
  confidence: Confidence;
  ambiguous: boolean;
}

export interface DetectedClaim {
  claim_type: string;
  category: string | null;
  scheme_or_service: ClaimField;
  department: ClaimField;
  requested_actions: string[];
  kb_entry_id?: string | null;
  amounts_inr?: number[];
}

export type FindingOutcome = "supported" | "contradicted" | "not_covered" | "ambiguous";

export interface KbFinding {
  aspect: string;
  outcome: FindingOutcome;
  detail_en: string;
  detail_ta: string | null;
  kb_entry_id?: string | null;
  source_ref?: string | null;
}

export interface KbSource {
  kb_entry_id: string;
  name: string;
  authority: string | null;
  official_url: string | null;
  source_citation?: string | null;
  last_reviewed?: string | null;
  source_url?: string | null;
  published?: string | null;
}

export interface GovernmentClaimResult {
  government_related: boolean;
  claim_status: GovernmentClaimStatus;
  claims: DetectedClaim[];
  matched_kb_entries?: string[];
  sources: KbSource[];
  findings: KbFinding[];
  limitations: string[];
  limitations_ta?: string[];
  safe_guidance_en?: string[];
  safe_guidance_ta?: string[];
  disclosure: string;
  disclosure_ta?: string;
}

export interface AnalyzeResponse {
  schema_version: string;
  analysis_id: string;
  risk: {
    assessment_status: "assessed" | "insufficient_content";
    level: RiskLevel;
    score: number;
    score_label: "indicator_strength";
    score_note: string;
    score_note_ta?: string;
    verdict_scope: string;
    verdict_scope_ta?: string | null;
  };
  evidence: EvidenceItem[];
  url_intelligence: {
    extracted_urls: string[];
    primary_url: string | null;
    domain_checks: { signal: string; confidence: Confidence; detail: string; url: string }[];
    provider_enabled: boolean;
    provider_result: ThreatIntelResult | null;
    privacy_note: string | null;
  };
  government_claim: GovernmentClaimResult;
  provenance: {
    content_source: ContentSource;
    image_origin: ImageOrigin | null;
    verification_status: "unverified";
    user_confirmed: boolean;
    attachment_received: boolean;
    ocr_status: "not_applicable" | "not_available_in_this_build" | "extracted" | "failed";
    character_count: number;
  };
  sender_assessment: {
    value_masked: string | null;
    kind: string;
    provenance: string;
    verification_status: "unverified";
    evidence_weight: "low";
    warnings: string[];
    warnings_ta?: string[];
  };
  missing_metadata: string[];
  limitations: string[];
  limitations_ta?: string[];
  safe_next_steps: string[];
  safe_next_steps_ta: string[];
  explanation: {
    en: string;
    ta: string;
    generated_by: "groq" | "gemini" | "template";
    ai_status?: AiStatus;
    model?: string | null;
    note: string | null;
    note_ta?: string | null;
  };
  provider_flags: {
    virustotal_enabled: boolean;
    virustotal_available: boolean | null;
    ai_provider: "groq" | "gemini" | "template";
    ai_enabled: boolean;
    ai_available: boolean | null;
  };
}
